### PR TITLE
chore: Improvements to metrics streaming endpoints [DET-4532]

### PR DESCRIPTION
**Pull request by mackrorysd**
## Description

Implementing a bunch of feedback that I got shortly after committing these calls. I'm wondering if our CI is going to flag this as incompatible, but anything disallowed by these changes was simply broken / wouldn't work / nonsensical before anyway. Furthermore, we tagged this API as "Internal" prior to release because a bigger change than this was being discussed (which we have decided against doing). These APIs are being implemented for a very specify use case, and as that use case progresses we might decide on other changes being necessary. The changes now are:

* Required parameters are now documented in the JSON schema
* Correct HTTP status code (400) are now returned for invalid input
* Correct HTTP status code (404) are now return for a non-existent experiment code
* The polling period is now an optional parameter in the query
* More efficient and comprehensive checking for experiment completion while polling

## Test Plan

Re-ran the integration tests, which should avoid an extra ~30 seconds of waiting on each request now! Also did some manual testing of negative cases checking HTTP codes, and various other sanity things.

### CODE REVIEW

File: proto/src/determined/api/v1/experiment.proto [hunk 1]
```diff
@@ -297,6 +307,9 @@ message MetricBatchesResponse {
 
 // Request metrics from all trials at a progress point of progress.
 message TrialsSnapshotRequest {
+  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
+    json_schema: { required: [ "experiment_id", "batches_processed", "metric_name", "metric_type" ] }
```
Comment (mackrorysd) [inline proto/src/determined/api/v1/experiment.proto:311]: @hamidzr In the generated Swagger UI locally, it only seems to be marking experiment_id as a required field, and not the others. I can't see what I've done wrong here - I think it was you that suggested this change - do you perhaps see anything?

Comment (hamidzr) [reply proto/src/determined/api/v1/experiment.proto:311]: could you share the generated swagger spec? Not sure why it would take one and not the others. When we generate the spec we want it to advertise these fields as camel case I saw something today that might suggest it's not applying the camelCase to these annotations as well. That might have a hand?

Comment (stoksc) [reply proto/src/determined/api/v1/experiment.proto:311]: I think experiment_id is marked as required implicitly because it is a path parameter. Compile without these annotations and it is still required. (according to `cat build/swagger/determined/api/v1/api.swagger.json | jq '.paths."/api/v1/experiments/{experimentId}/metrics-stream/trials-snapshot"'`). Peeking at create experiment looks like it's not generating quite what is expected either.

File: proto/src/determined/api/v1/experiment.proto [hunk 2]
```diff
@@ -297,6 +307,9 @@ message MetricBatchesResponse {
...
+  };
   // The id of the experiment.
   int32 experiment_id = 1;
   // The point of progress at which to query metrics.
```
Comment (stoksc) [inline proto/src/determined/api/v1/experiment.proto:315]: on the line below
`int32 batches_processed = 2 [(grpc.gateway.protoc_gen_swagger.options.openapiv2_field) = {required: ['batches_processed']}];`
seems to produce the expected result. found it just ignoring what we're doing and trying to figure out how to do it independently.

Comment (stoksc) [reply proto/src/determined/api/v1/experiment.proto:315]: I see what we're doing documented though, not sure why it doesn't work.

Comment (stoksc) [reply proto/src/determined/api/v1/experiment.proto:315]: https://github.com/grpc-ecosystem/grpc-gateway/issues/665 ok that explains it.

**Review by stoksc — APPROVED** [review]
everything lgtm, of course except for the grpc gateway required annotations, but we can probably check/fix all of them in one go with another ticket.

File: proto/src/determined/api/v1/experiment.proto (hunk 2 above)
Comment (mackrorysd) [reply proto/src/determined/api/v1/experiment.proto:306]: Yep that appears to be working to me. Going with that!

**Review by stoksc — APPROVED** [review]

File: proto/src/determined/api/v1/experiment.proto (hunk 1 above)
Comment (hamidzr) [reply proto/src/determined/api/v1/experiment.proto]: this should address that https://github.com/determined-ai/determined/pull/1681
